include/deal.II/base/mpi.templates.h

```diff
@@ -186,6 +186,129 @@ namespace Utilities
               std::copy(values.begin(), values.end(), output.begin());
           }
       }
+
+      template<typename T>
+      std::map<unsigned int, std::vector<T> >
+      send_and_receive(const MPI_Comm                                         &comm,
+                       const std::map<unsigned int, typename std::vector<T> > &objects_to_send)
+      {
+#ifndef DEAL_II_WITH_MPI
+        (void)comm;
+        (void)objects_to_send;
+        Assert (false, ExcMessage ("The function send_and_receive doesn't make"
+                                   "any sense if you do not have MPI enabled. "));
+#else
+        auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+        auto my_proc = dealii::Utilities::MPI::this_mpi_process(comm);
+
+        // Collective communication telling each process
+        // how many objects I'm sending: this value shall be contained
+        // inside send_amount
+        std::vector<unsigned int> send_amount(n_procs, 0);
+        // Vector to store, in each entry, the number of objects the corresponding
+        // process shall send to this process
+        std::vector<unsigned int> rec_amount(n_procs, 0);
+
+        for (const auto &m: objects_to_send)
+          send_amount[m.first] = m.second.size();
+
+        MPI_Alltoall(
+          &send_amount[0], 1, MPI_UNSIGNED,
+          &rec_amount[0], 1, MPI_UNSIGNED, comm);
+        unsigned int number_p2p_comm = 0;
+        for (unsigned int rank=0; rank< n_procs; ++rank)
+          if (rec_amount[rank] != 0)
+            ++number_p2p_comm;
+        number_p2p_comm += objects_to_send.size();
+        // Creating the MPI data type to send the object
+        unsigned int size_of_T = sizeof(T);
+        MPI_Datatype ptype;
+        MPI_Type_contiguous(size_of_T,MPI_CHAR,&ptype);
+        MPI_Type_commit(&ptype);
+        std::map<unsigned int, std::vector<T> > received_objects;
+        std::vector<MPI_Request> requests(number_p2p_comm);
+        unsigned int req_n = 0; // index running on the request vector
+        // Sending part
+        for (const auto &rank_vec : objects_to_send)
+          {
+            const auto &rank = rank_vec.first;
+            const auto &vec  = rank_vec.second;
+            MPI_Isend(&(vec[0]), vec.size(), ptype,
+                      rank, n_procs*rank+my_proc, comm, &requests[req_n++]);
+          }
+
+        // Receiving part
+        for (unsigned int rank=0; rank < n_procs; ++rank)
+          if (rec_amount[rank] != 0)
+            {
+              received_objects[rank].resize(rec_amount[rank]);
+              auto &vec = received_objects[rank];
+              MPI_Irecv(&(vec[0]), rec_amount[rank], ptype,
+                        rank, n_procs*my_proc+rank, comm, &requests[req_n++]);
```

Comment:
This is inefficient: you are working through the ranks one by one. But if a message from the rank you're currently looking at hasn't come in, you have to wait. The better way is to work on messages in the order as they come in, until you've gotten all you were expecting to get.

Reply:
`GridTools::exchange_cell_data_to_ghosts()` shows how to do that.

Reply:
These are non-blocking communications. No waiting is involved. The function call returns immediately. Why do you think this would be inefficient?

Reply:
actually, `exchange_cell_data_to_ghosts` does exactly the same as this function...

Reply:
Yes, these are nonblocking sends/recvs, so this is fine @bangerth

Reply:
But as another note: Setting up the buffer of MPI_IRecv first and then doing the ISend should speed this communication up, at least that is the usual example code (on an unrelated question, but still: https://stackoverflow.com/questions/9710650/mpi-isend-irecv-performance-behavior), and what is claimed on the last slide here: http://www.cas.mcmaster.ca/%7Enedialk/COURSES/mpi/Lectures/lec2_1.pdf
I have not measured that though, feel free to correct me if that is not the case.